File: Assets/Game/Scripts/Moduls/P_Modul_03.lua
local ObjectStateBackupUtils = Ballance2.Utils.ObjectStateBackupUtils

---@class P_Modul_03 : ModulBase
---@field P_Modul_03_Floor PhysicsObject
---@field P_Modul_03_Gate PhysicsObject
---@field P_Modul_03_Wall01 PhysicsObject
---@field P_Modul_03_Wall02 PhysicsObject
---@field P_Modul_03_Wall03 PhysicsObject
---@field P_Modul_03_Wall04 PhysicsObject
---@field P_Modul_03_Wall05 PhysicsObject
---@field P_Modul_03_Wall06 PhysicsObject
---@field P_Modul_03_Wall07 PhysicsObject
P_Modul_03 = ModulBase:extend()

function P_Modul_03:new()
  P_Modul_03.super.new(self)
  self.EnableBallRangeChecker = true
  self.BallCheckeRange = 30
end

function P_Modul_03:Active()
  ModulBase.Active(self)
  self.P_Modul_03_Floor:Physicalize()
  self.P_Modul_03_Gate:Physicalize()
  self.P_Modul_03_Wall01:Physicalize()
  self.P_Modul_03_Wall02:Physicalize()
  self.P_Modul_03_Wall03:Physicalize()
  self.P_Modul_03_Wall04:Physicalize()
  self.P_Modul_03_Wall05:Physicalize()
  self.P_Modul_03_Wall06:Physicalize()
  self.P_Modul_03_Wall07:Physicalize()
  self.P_Modul_03_Floor.CollisionID = GamePlay.BallSoundManager:GetSoundCollIDByName('Wood')
end
function P_Modul_03:Deactive()
  self.P_Modul_03_Floor:UnPhysicalize(true)
  self.P_Modul_03_Gate:UnPhysicalize(true)
  self.P_Modul_03_Wall01:UnPhysicalize(true)
  self.P_Modul_03_Wall02:UnPhysicalize(true)
  self.P_Modul_03_Wall03:UnPhysicalize(true)
  self.P_Modul_03_Wall04:UnPhysicalize(true)
  self.P_Modul_03_Wall05:UnPhysicalize(true)
  self.P_Modul_03_Wall06:UnPhysicalize(true)
  self.P_Modul_03_Wall07:UnPhysicalize(true)
  ModulBase.Deactive(self)
end
function P_Modul_03:Reset()
  ObjectStateBackupUtils.RestoreObjectAndChilds(self.gameObject)
end
function P_Modul_03:Backup()
  ObjectStateBackupUtils.BackUpObjectAndChilds(self.gameObject)
end
function P_Modul_03:BallEnterRange()
  if self.IsActive then
    self.P_Modul_03_Floor:WakeUp()
  end
end

function CreateClass:P_Modul_03()
  return P_Modul_03()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Modul_03.cs
using Ballance2.Utils;
using BallancePhysics.Wapper;

namespace Ballance2.Game.GamePlay.Moduls
{
  public class P_Modul_03 : ModulBase 
  {
    public PhysicsObject P_Modul_03_Floor;
    public PhysicsObject P_Modul_03_Gate;
    public PhysicsObject[] P_Modul_03_Wall = new PhysicsObject[6];

    public P_Modul_03() {
      EnableBallRangeChecker = true;
      BallCheckeRange = 30;
    }

    public override void Active()
    {
      base.Active();
      P_Modul_03_Floor.gameObject.SetActive(true);
      P_Modul_03_Gate.gameObject.SetActive(true);
      P_Modul_03_Floor.CollisionID = GamePlayManager.Instance.BallSoundManager.GetSoundCollIDByName("Wood");
      P_Modul_03_Floor.Physicalize();
      P_Modul_03_Gate.Physicalize();
 
      foreach(var wall in P_Modul_03_Wall) {
        wall.gameObject.SetActive(true);
        wall.Physicalize();
      }

    }

    public override void Deactive()
    {
      P_Modul_03_Floor.UnPhysicalize();
      P_Modul_03_Gate.UnPhysicalize();
      foreach(var wall in P_Modul_03_Wall)
        wall.UnPhysicalize();
      base.Deactive();
    }

    public override void ActiveForPreview()
    {
      gameObject.SetActive(true);
    }

    public override void DeactiveForPreview()
    {
      gameObject.SetActive(false);
    }

    public override void Reset(ModulBaseResetType type)
    {
      ObjectStateBackupUtils.RestoreObjectAndChilds(gameObject);
    }

    public override void Backup()
    {
      ObjectStateBackupUtils.BackUpObjectAndChilds(gameObject);
    }

    protected override void BallEnterRange()
    {
      if (!IsPreviewMode && IsActive)
        P_Modul_03_Floor.WakeUp();
    }
  }
}